File: ruby/small-problems/1000-lights.rb
# You have a bank of switches before you numbered from 1 to 1000. Each switch is
# connected to exactly one light that is initially off. You walk down the row of
# switches, and turn every one of them on. Then, you go back to the beginning
# and toggle switches 2, 4, 6, and so on. Repeat this for switches 3, 6, 9, and
# so on, and keep going until you have been through 1000 repetitions of this
# process.
#
# Write a program that determines how many lights are on at the end, and which
# lights are on.
#
# Example with 5 lights:
#
# round 1: every light is turned on
# round 2: lights 2 and 4 are now off; 1, 3, 5 are on
# round 3: lights 2, 3, and 4 are now off; 1 and 5 are on
# round 4: lignts 2 and 3 are now off; 1, 4, and 5 are on
# round 5: lights 2, 3, and 5 are now off; 1 and 4 are on
# The result is that 2 lights are left on, lights 1 and 4.
#
# With 10 lights, 3 lights are left on: lights 1, 4, and 9.

TOGGLE_CYCLES = 10

def toggle_lights(lights)
  1.upto(lights.size) do |cycle|
    lights.each do |key, value|
      lights[key] = toggle_switch(value) if key >= cycle && key % cycle == 0
    end
  end
end

def toggle_switch(light)
  light == "on" ? "off" : "on"
end

def lights_on(lights)
  lights.keys.select { |key| lights[key] == "on" }
end

def lights_off(lights)
  lights.keys.select { |key| lights[key] == "off" }
end

def display_lights_on(lights)
  puts "Lights ON:  #{lights_on(lights).size} - #{lights_on(lights).join(", ")}"
  puts "Lights OFF: #{lights_off(lights).size} - #{lights_off(lights).join(", ")}"
end

lights = 1.upto(TOGGLE_CYCLES).each.with_object({}) do |index, hash|
  hash[index] = "off"
end

toggle_lights(lights)
display_lights_on(lights)

File: javascript/small_problems/1000_lights.js
// You have a bank of switches before you, numbered from 1 to n. Each switch is
// connected to exactly one light that is initially off. You walk down the row
// of switches, and toggle every one of them. You go back to the beginning, and
// on this second pass you toggle switches 2, 4, 6, and so on. On the third
// pass, you go back again to the beginning and toggle switches 3, 6, 9, and so
// on. You repeat this process and keep going until you have been through n
// repetitions.
//
// Write a program that takes one argument, total number of switches, and
// returns an array of lights that are on after n repetitions.
//
// detailed result of each round for 5 lights
// Round 1: all lights are on
// Round 2: lights 2 and 4 are now off; 1, 3, and 5 are on
// Round 3: lights 2, 3, and 4 are now off; 1 and 5 are on
// Round 4: lights 2 and 3 are now off; 1, 4, and 5 are on
// Round 5: lights 2, 3, and 5 are now off; 1 and 4 are on


// INPUT
// a number representing the number of switches and iterations

// OUTPUT
// an array where each element represent the number of a light ON after all
// the iterations

// REQUIREMENTS
// – all lights are OFF at the beginning
// – we iterate the same times as the number passed in to the function
//   – in the 1st iteration all lights are set to ON
//   – on each following iteration we toggle the switch equivalent to the number
//     iteration and the following switches multiples of that number.
//     (a switch ON pass to be OFF and viceversa)

// ALGORITHM
// – create an array of length equal to the number passed in to the function
//   – set all lights to OFF
// – iterate trough the array
//   – toggle lights that are multiple of the number of the iteration
// – filter array with lights ON and return it

function lightsOn(n) {
  var result = [];
  var lights = setLights(n);

  for (var i = 1; i <= n; i++) {
    lights = toggleLights(lights, i);
  }

  lights.forEach(function(light, index) {
    if (light) {
      result.push(index + 1);
    }
  });

  return result;
}

function setLights(n) {
  var lights = [];

  for (var i = 0; i < n; i++) {
    lights.push(false);
  }

  return lights;
}

function toggleLights(lights, round) {
  return lights.map(function(light, index) {
    return (index + 1) % round === 0 ? !light : light;
  });
}

lightsOn(1);    // [1]
lightsOn(2);    // [1]
lightsOn(5);    // [1, 4]
lightsOn(25);   // [1, 4, 9, 16, 25]
lightsOn(100);  // [1, 4, 9, 16, 25, 36, 49, 64, 81, 100]
